File: go/graph/component.go
package graph

// 图的连通分量
type Component struct {
	graph Graph
	visited []bool
	// 连通分量
	ccount int
	// 并查集
	id []int
}

func NewComponent(graph Graph) *Component {
	comp := &Component{}
	visited := make([]bool, graph.V())
	comp.graph = graph
	comp.visited = visited
	comp.ccount = 0
	comp.id = make([]int, graph.V())

	for i := 0; i < len(comp.id); i++ {
		comp.id[i] = -1
	}

	for i := 0; i < graph.V(); i++ {
		if !comp.visited[i] {
			comp.dfs(i);
			comp.ccount++;
		}
	}


	return comp
}

// 深度优先遍历
func (this *Component) dfs(v int) {
	this.visited[v] = true
	this.id[v] = this.ccount
	vertexs, _ := this.graph.Vertexs(v)
	for i := 0; i < len(vertexs); i++ {
		if !this.visited[vertexs[i]] {
			this.dfs(vertexs[i])
		}
	}
}

func (this *Component) Count() int {
	return this.ccount
}

func (this *Component) IsConnected(v, w int) bool {
	if v < 0 || v >= len(this.id) || w < 0 || w >= len(this.id) {
		return false
	}
	return this.id[v] == this.id[w]
}

File: js/graphMST/component.js
// 连通分量
class Component {
  constructor(graph) {
    const vCount = graph.V();
    this.graph = graph;
    this.visited = Array(vCount);
    this.ccount = 0;
    this.id = Array(vCount);

    for (let i = 0; i < vCount; i++) {
      this.visited[i] = false;
    }

    for (let i = 0; i < vCount; i++) {
      if (!this.visited[i]) {
        this.dfs(i);
        this.ccount++;
      }
    }
  }

  dfs(v) {
    this.visited[v] = true;
    this.id = this.ccount;
    adjs = this.graph.Adj(v);

    for (let i = 0; i < adjs.length; i++) {
      const w = adjs.Other(v);
      if (!this.visited[w]) {
        this.dfs(w);
      }
    }
  }

  /**
   * 顶点 v 和顶点 w 是否是连通的
   * @param {int} v
   * @param {int} w
   * @returns {bool}
   */
  IsConnected(v, w) {
    if (v < 0 || v >= this.graph.V() || w < 0 || w >= this.graph.V()) {
      return false;
    }

    return this.id[v] === this.id[w];
  }

  // 图的连通分量个数
  Count() {
    return this.count;
  }
}

module.exports = Component;
